scripts/train/train.py

```diff
@@ -164,170 +165,315 @@ def build_dataloader(cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
             tokenizer,
             device_batch_size,
         )
-
     else:
         raise ValueError(f'Not sure how to build dataloader with config: {cfg}')
 
 
 def main(cfg: DictConfig):
-    # Check for incompatibilities between the model and data loaders
-    validate_config(cfg)
-
     # Filter deprecation warning from torch internal usage
     warnings.filterwarnings(
         action='ignore',
         category=UserWarning,
         message=
-        f'torch.distributed.*_base is a private function and will be deprecated.*'
+        'torch.distributed.*_base is a private function and will be deprecated.*'
     )
 
-    cfg.dist_timeout = cfg.get('dist_timeout', 600.0)
+    # Check for incompatibilities between the model and data loaders
+    validate_config(cfg)
 
-    reproducibility.seed_all(cfg.seed)
-    dist.initialize_dist(get_device(None), timeout=cfg.dist_timeout)
+    # Resolve all interpolation variables as early as possible
+    om.resolve(cfg)
 
-    # Run Name
-    if cfg.get('run_name') is None:
-        cfg.run_name = os.environ.get('RUN_NAME', 'llm')
+    # Set seed first
+    seed: int = pop_config(cfg, 'seed', must_exist=True)
+    reproducibility.seed_all(seed)
 
-    # Get batch size info
+    # Initialize distributed setting with seed
+    dist_timeout: Union[int, float] = pop_config(cfg,
+                                                 'dist_timeout',
+                                                 must_exist=False,
+                                                 default_value=600.0)
+    dist.initialize_dist(get_device(None), timeout=dist_timeout)
+
+    # Get global and device batch size information from distributed/single node setting
     cfg = update_batch_size_info(cfg)
 
-    # Read FSDP Config as a dict
-    fsdp_config = cfg.get('fsdp_config', None)
-    fsdp_config = om.to_container(fsdp_config,
-                                  resolve=True) if fsdp_config else None
-    assert isinstance(fsdp_config, Dict) or fsdp_config is None
+    print('Logging config...')
+    log_config(cfg)
```

Comment:
The log config call actually needs to happen _after_ the trainer is created, because creating the trainer is what initializes wandb and mlflow.

Reply:
The `cfg` will be empty after the `trainer` has been initialized since we are popping out from it to check for improperly formatted variables. 

A slightly verbose solution is to create a `logged_cfg` variable that we pop all variables from `cfg` to logged_cfg to solely for logging purposes.

Reply:
Ah, good point. Should we make a deepcopy at the start and log that at the end?

Reply:
ya let's do that